native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java

```diff
@@ -2,84 +2,185 @@
+                        Thread.currentThread().getId() +
+                        " has bind to wrong cores");
             }
         }
 
-        setAffinity0(sets);
+        bindToCores(this.coreList); // reset to default
+
+        for (int key : affinity) {
+            availableCores.get(key).remove(Thread.currentThread().getId());
+        }
     }
 
     public synchronized void setOmpAffinity() {
-        setOmpAffinity0(getAffinity());
+        // TODO omp thread can be managed availableCores.
+        int[] ret = setOmpAffinity0(coreList);
+
+        boolean isAllSuccess = true;
+
+        for (int r : ret) {
+            if (r != 0) {
+                isAllSuccess = false;
+                break;
+            }
+        }
+
+        if (!isAllSuccess) {
+            throw new BindException(coreList);
+        }
     }
 
     public synchronized int[] getAffinity() {
-        int[] temp = new int[cpuCounts];
+        int[] temp = new int[coreCounts];
         Arrays.fill(temp, 0);
-        getAffinity0(temp, cpuCounts);
 
-        int rIndex = 0;
-        int[] ret = new int[getSetted(temp)];
+        if (getAffinity0(temp) != 0) {
+            throw new RuntimeException("failed to get the affinity info of thread " +
+                    Thread.currentThread().getName() +
+                    " #id " + Thread.currentThread().getId());
+        }
+
+        return humanReadable(temp);
+    }
+
+    public synchronized int[] getOmpAffinity() {
+        int[] ret = getOmpAffinity0(this.coreCounts);
 
-        for (int tIndex = 0; tIndex < cpuCounts; tIndex++) {
-            int elem = temp[tIndex];
+        boolean isAllSucc = true;
 
-            if (elem == 1) {
-                ret[rIndex] = tIndex;
-                rIndex++;
+        for (int i = 0; i < ret.length; i++) {
+            if (ret[i] == -1) {
+                isAllSucc = false;
+                break;
             }
         }
 
-        for (int elem : ret) {
-            availableCPUs[elem] = false;
+        if (!isAllSucc) {
+            throw new RuntimeException("failed to get the affinity info of thread " +
+                    Thread.currentThread().getName() +
+                    " #id " + Thread.currentThread().getId());
         }
 
         return ret;
     }
 
-    private int getSetted(int[] sets) {
-        int counts  = 0;
-        for (int set : sets) {
-            if (set == 1) {
-                counts++;
+    public Map<Integer, List<Long>> stats() {
+        return Collections.unmodifiableMap(availableCores);
+    }
+
+    private int[] humanReadable(int[] set) {
+        int counts = 0;
+        List<Integer> temp = new ArrayList<Integer>();
+
+        for (int i = 0; i < set.length; i++) {
+            if (set[i] == 1) {
+                temp.add(i);
             }
         }
 
-        return counts;
+        int[] result = new int[temp.size()];
+        for (int i = 0; i < temp.size(); i++) {
```

Comment:
just `temp.toArray()` ? and then rename temp to `result`

Reply:
`toArray` will return an `Object` array, which is not the same as `[]`.